File: longestwordindic.java

time complexity: O(n)
space complexity: O(1)

class Solution {
    private static class Node{
		Node[] childs;
		boolean isLeaf;
		public Node(){
			this.childs = new Node[26];
		}
	}
	private void insert(Node node, String s){
		char[] arr = s.toCharArray();
		if(arr.length == 1){
			node.childs[arr[0] - 'a'] = new Node();
			if(len < 1){
				this.res = s;
				this.len = 1;
			}
			node = node.childs[arr[0] - 'a'];
			node.isLeaf = true;
		}else{
			for(int i = 0; i < arr.length - 1; i++){
				int c = arr[i] - 'a';
				if(node.childs[c] == null){
					return;
				}
				node = node.childs[c];
			}
			if(node.isLeaf){
				node.childs[arr[arr.length - 1] - 'a'] = new Node();
				node = node.childs[arr[arr.length - 1] - 'a'];
				node.isLeaf = true;
				if(arr.length > len){
					this.res = s;
					len = arr.length;
				}
			}
		}
	}
	private String res = "";
	private int len = 0;
    
    public String longestWord(String[] words) {
        Arrays.sort(words);
		Node root = new Node();
		for(String word : words){
			insert(root, word);
		}
		return res;
        
    }
}

File: LongestWordInDic.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.end_of_word = False
        self.word = ""
class Trie:
    def __init__(self):
        self.root = TrieNode()
        
    # Time Complexity: O(len of words * avg len of word)    
    def insert(self, word: str) -> None:
        curr = self.root
        
        for c in word:
            if c not in curr.children:
                curr.children[c] = TrieNode()
            curr = curr.children[c]
        curr.end_of_word = True
        curr.word = word
        
    # Time Complexity: O(len of words * avg len of word)
    def bfs(self) -> str:
        
        queue = collections.deque()
        
        queue.append(self.root)
        result = ""
        
        while queue:
            curr = queue.popleft()
            for n in curr.children.values():
                if n.end_of_word:
                    queue.append(n)

                    if len(n.word) > len(result) or n.word < result:
                        result = n.word
                    
        return result
        
class Solution:
    def longestWord(self, words: List[str]) -> str:
        trie = Trie()
        for word in words:
            trie.insert(word)
            
        return trie.bfs()
    
    # Time Complexity: O(len of words * avg len of word) actually it will be O(2(len of words * avg len of word)) for inserting and bfs both
    # Space Complexity: O(len of words * avg len of word) because both trie and queue takes the same number of nodes